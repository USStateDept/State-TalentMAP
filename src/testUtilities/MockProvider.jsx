import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { FlagsProvider } from 'flag';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import getFlags from '../flags';

const flags = () => getFlags();
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

export const MockTestProvider = ({ flagProps, providerProps, memoryProps, children }) => (
  <FlagsProvider flags={flags()} {...flagProps}>
    <Provider store={mockStore({})} {...providerProps}>
      <MemoryRouter {...memoryProps}>
        {children}
      </MemoryRouter>
    </Provider>
  </FlagsProvider>
);

MockTestProvider.propTypes = {
  flagProps: PropTypes.shape({}),
  providerProps: PropTypes.shape({}),
  memoryProps: PropTypes.shape({}),
  children: PropTypes.node.isRequired,
};

MockTestProvider.defaultProps = {
  flagProps: {},
  providerProps: {},
  memoryProps: {},
};

export default MockTestProvider;
