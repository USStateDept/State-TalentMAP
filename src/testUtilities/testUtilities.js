import sinon from 'sinon';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MockAdapter from 'axios-mock-adapter';
import api from '../api';

// Tests all functions within a mapDispatchToProps function.
// We take two arguments, one is the entire mapDispatchToProps function, which
// can be exported from the Container/Component. The second is a config
// object with any parameters we want to pass to the function with the same name.
// For example, to pass the argument "test" to the getData() function, and "test"
// and {1} to submitData() function, we could create an object such as
// { getData: ["test"], submitData: ["test", 1] }
// Then we spy dispatch and call each property of the mapDispatchToProps
// object. Calling each property should trigger the call count on the spy,
// which we test. We reset call count between tests since we're using the
// calledOnce method.
export function testDispatchFunctions(mapDispatchToProps, config = {}) {
  it('should call all mapDispatchToProps actions', () => {
    // create spy
    const dispatchSpy = sinon.spy();
    // object with all functions
    const spiedProps = mapDispatchToProps(dispatchSpy);
    // iterate on each key
    Object.keys(spiedProps).forEach((prop) => {
      // set the prop to test
      const propToTest = spiedProps[prop];
      // set any params passed in the config object
      let params = [];
      // if there any arguments for that param, set them
      if (config[prop]) { params = config[prop]; }
      // call the prop
      propToTest(...params);
      // ensure it was called once
      sinon.assert.calledOnce(dispatchSpy);
      // reset so that called count is 0
      dispatchSpy.reset();
    });
  });
}

export function setupAsyncMocks() {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const mockAdapter = new MockAdapter(api());

  return { mockStore, mockAdapter };
}

// Form a mock adapter using a url, response, type, and callbackType.
// Spy on the mock and return the mock and the spy.
export const spyMockAdapter = ({ url = '', response = [], type = 'onGet', callbackType = 'reply' }) => {
  const mockAdapter = new MockAdapter(api());
  const spy = sinon.spy(() => response);
  const mock = () => mockAdapter[type](url)[callbackType](spy);
  return { mock, spy };
};

// Expect that some assertion was performed on a provided spy.
// Optional callback once complete.
export const expectMockWasCalled = ({ spy = () => {}, cb = () => {}, timeout = 0, assertion = 'calledOnce' }) => {
  setTimeout(() => {
    sinon.assert[assertion](spy);
    cb();
  }, timeout);
};
