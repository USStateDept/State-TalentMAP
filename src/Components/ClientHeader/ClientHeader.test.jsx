import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import TestUtils from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import sinon from 'sinon';
import { includes } from 'lodash';
import ClientHeaderContainer, { ClientHeader, mapDispatchToProps } from './ClientHeader';
import { testDispatchFunctions } from '../../testUtilities/testUtilities';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('ClientHeader', () => {
  let props;
  beforeEach(() => {
    props = {
      id: 1,
      unset: () => {},
      history: { listen: () => {} },
      client: { perdet_seq_number: 1, name: 'Mary' },
    };
  });

  it('is defined', () => {
    const wrapper = shallow(<ClientHeader {...props} />);
    expect(wrapper).toBeDefined();
  });

  it("displays a proxied CDO's name", () => {
    const wrapper = shallow(<ClientHeader {...props} bidderPortfolioSelectedCDO={{ name: 'Andrew Proxy' }} />);
    expect(wrapper.find('#search-as-name').text()).toMatch(/Proxying as Andrew Proxy/);
    expect(wrapper).toBeDefined();
  });

  it('uses different colors when proxied as another CDO', () => {
    const wrapper = shallow(<ClientHeader {...props} bidderPortfolioSelectedCDO={{ name: 'Andrew Proxy' }} />);
    expect(wrapper.find('SkeletonTheme').at(0).props().color).toBe('#FFF1D2');
    wrapper.setProps({ ...props, bidderPortfolioSelectedCDO: { name: 'Self', isCurrentUser: true } });
    wrapper.update();
    expect(wrapper.find('SkeletonTheme').at(0).props().color).toBe('#DCE4EF');
    expect(wrapper).toBeDefined();
  });

  it('is displays the proxy CDO when is exists and is not the current user', () => {
    const wrapper = shallow(<ClientHeader {...props} bidderPortfolioSelectedCDO={{ name: 'Mike', isCurrentUser: false }} />);
    const text = wrapper.find('#search-as-name').text();
    expect(includes(text, '(Proxying as Mike)')).toBe(true);
  });

  it('calls unset() on click', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<ClientHeader
      {...props}
      unset={spy}
      client={{ perdet_seq_number: 1 }}
    />);
    wrapper.find('button').simulate('click');
    sinon.assert.calledOnce(spy);
  });

  it('it mounts', () => {
    const wrapper = TestUtils.renderIntoDocument(
      <Provider store={mockStore({ clientView: { client: {} } })}>
        <MemoryRouter>
          <ClientHeaderContainer {...props} />
        </MemoryRouter>
      </Provider>);
    expect(wrapper).toBeDefined();
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<ClientHeader {...props} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('matches snapshot when proxying as another CDO', () => {
    const wrapper = shallow(<ClientHeader {...props} bidderPortfolioSelectedCDO={{ name: 'Andrew Proxy' }} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});

describe('mapDispatchToProps', () => {
  testDispatchFunctions(mapDispatchToProps);
});
