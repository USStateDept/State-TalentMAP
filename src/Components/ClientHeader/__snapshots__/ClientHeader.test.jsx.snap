// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientHeader matches snapshot 1`] = `
<div
  className="client-header-container"
  id="client-header"
  style={Object {}}
>
  <div
    className="usa-banner client-header  "
  >
    <div
      className="usa-grid usa-banner-inner"
    >
      <div
        className=""
      >
        <SkeletonTheme
          color="#DCE4EF"
          highlightColor="#8BA6CA"
        >
          <Link
            replace={false}
            to="/profile/public/1"
          >
            <FontAwesome
              name="chevron-left"
            />
            <span>
              Client Dashboard
            </span>
          </Link>
        </SkeletonTheme>
      </div>
      <div>
        <SkeletonTheme
          color="#DCE4EF"
          highlightColor="#8BA6CA"
        >
          <span>
            <FontAwesome
              name="clipboard"
            />
            <span
              id="search-as-name"
            >
              Position Search for 
              Mary
            </span>
          </span>
        </SkeletonTheme>
      </div>
      <div>
        <SkeletonTheme
          color="#DCE4EF"
          highlightColor="#8BA6CA"
        >
          <button
            className="unstyled-button"
            onClick={[Function]}
          >
            <FontAwesome
              name="close"
            />
            <span>
              Exit client view
            </span>
          </button>
        </SkeletonTheme>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClientHeader matches snapshot when proxying as another CDO 1`] = `
<div
  className="client-header-container"
  id="client-header"
  style={Object {}}
>
  <div
    className="usa-banner client-header client-header--alternate "
  >
    <div
      className="usa-grid usa-banner-inner"
    >
      <div
        className=""
      >
        <SkeletonTheme
          color="#FFF1D2"
          highlightColor="#FAD980"
        >
          <Link
            replace={false}
            to="/profile/public/1"
          >
            <FontAwesome
              name="chevron-left"
            />
            <span>
              Client Dashboard
            </span>
          </Link>
        </SkeletonTheme>
      </div>
      <div>
        <SkeletonTheme
          color="#FFF1D2"
          highlightColor="#FAD980"
        >
          <span>
            <FontAwesome
              name="clipboard"
            />
            <span
              id="search-as-name"
            >
              Position Search for 
              Mary
               (Proxying as Andrew Proxy)
            </span>
          </span>
        </SkeletonTheme>
      </div>
      <div>
        <SkeletonTheme
          color="#FFF1D2"
          highlightColor="#FAD980"
        >
          <button
            className="unstyled-button"
            onClick={[Function]}
          >
            <FontAwesome
              name="close"
            />
            <span>
              Exit client view
            </span>
          </button>
        </SkeletonTheme>
      </div>
    </div>
  </div>
</div>
`;
